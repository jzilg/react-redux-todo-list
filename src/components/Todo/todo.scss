@import "../../style/color";

.todo {
    display: grid;
    grid-template-columns: 1.6fr 1fr 2fr 1fr .4fr;
    grid-column-gap: 10px;
    align-items: center;
    height: 100%;
}

.input {
    height: 100%;
    border: 0;
    background-color: transparent;
    text-overflow: ellipsis;
}

.last-event-input {
    margin-left: 4px;
    @extend .input;
}

.schedule-input {
    width: 42px;
    margin-right: 5px;
    margin-left: 5px;
    text-align: center;
    @extend .input;
}

.delete-btn {
    height: 100%;
    border-style: none;
    background-color: transparent;
    font-size: 24px;
    color: $grey;
    transition: background-color .2s, color .2s;

    &:hover {
        background-color: $red;
        color: #fff;
    }
}
