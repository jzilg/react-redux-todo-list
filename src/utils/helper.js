import moment from 'moment'

/**
 * @returns {string} todays date (YYYY-MM-DD)
 */
export function getTodaysDate() {
    return moment().format('YYYY-MM-DD')
}

/**
 * @param {string} startDate moment date format
 * @param {string} endDate moment date format
 * @returns {number} duration in days
 */
export function getDurationBetweenDates(startDate, endDate) {
    const startDateMoment = moment(startDate)
    const endDateMoment = moment(endDate)
    return moment.duration(startDateMoment.diff(endDateMoment)).asDays()
}
