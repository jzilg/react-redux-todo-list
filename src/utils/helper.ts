import moment from 'moment'

/**
 * @returns {string} todays date (YYYY-MM-DD)
 */
export function getTodaysDate(): string {
    return moment().format('YYYY-MM-DD')
}

/**
 * @param {string} startDate moment date format
 * @param {string} endDate moment date format
 * @returns {number} duration in days
 */
export function getDurationBetweenDates(startDate, endDate): number {
    const startDateMoment = moment(startDate)
    const endDateMoment = moment(endDate)
    return startDateMoment.diff(endDateMoment, 'days')
}

function randomIntFromInterval(min: number, max: number): number {
    const { random, floor } = Math
    return floor(random() * (max - min + 1) + min)
}

/**
 * @returns {number} new id base64 of timestamp and random number
 */
export function createUniqueId(): number {
    const date = new Date()
    const time = date.getTime()
    const randomNumber = randomIntFromInterval(10000, 99999)
    return Math.floor(time / randomNumber)
}
