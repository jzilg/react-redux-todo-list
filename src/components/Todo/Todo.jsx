import React from 'react'
import PropTypes from 'prop-types'
import Urgency from '../Urgency'
import './todo.scss'

class Todo extends React.Component {
    constructor(props) {
        super()

        const { todo } = props

        this.state = {
            id: todo.id,
            name: todo.name,
            schedule: parseInt(todo.schedule, 10) || 1,
            lastEvent: todo.lastEvent || props.today,
        }

        this.inputChange = this.inputChange.bind(this)
        this.saveTodo = this.saveTodo.bind(this)
        this.removeTodo = this.removeTodo.bind(this)
        this.setInputToday = this.setInputToday.bind(this)
    }

    setInputToday() {
        const { state, props } = this
        if (state.lastEvent === props.today) {
            return
        }

        this.setState({
            lastEvent: props.today,
        }, () => {
            this.saveTodo()
        })
    }

    inputChange(event) {
        const { name } = event.target
        let { value } = event.target

        if (name === 'schedule') {
            value = parseInt(value, 10)
        }

        this.setState({
            [name]: value,
        })
    }

    saveTodo() {
        const { saveTodo } = this.props
        saveTodo(this.state)
    }

    removeTodo() {
        const { removeTodo } = this.props
        removeTodo(this.state)
    }

    render() {
        const { isLoading, today } = this.props
        const {
            id,
            name,
            schedule,
            lastEvent,
        } = this.state

        return (
            <form key={id} styleName="todo" title={name}>
                <input
                    type="text"
                    name="name"
                    value={name}
                    placeholder="Name"
                    autoComplete="off"
                    onChange={this.inputChange}
                    onBlur={this.saveTodo}
                    autoFocus={!name.length}
                    styleName="input"
                />
                <div>
                    <span>
                        Every
                    </span>
                    <input
                        type="number"
                        name="schedule"
                        styleName="schedule-input"
                        value={schedule}
                        min="1"
                        autoComplete="off"
                        onChange={this.inputChange}
                        onBlur={this.saveTodo}
                    />
                    <span>
                        {schedule === 1 ? 'Day' : 'Days'}
                    </span>
                </div>
                <div>
                    <span>
                        Last done on
                    </span>
                    <input
                        type="date"
                        name="lastEvent"
                        styleName="last-event-input"
                        value={lastEvent}
                        onChange={this.inputChange}
                        onBlur={this.saveTodo}
                    />
                    <button
                        type="button"
                        onClick={this.setInputToday}
                    >
                        Today
                    </button>
                </div>
                <Urgency
                    today={today}
                    lastEvent={lastEvent}
                    schedule={schedule}
                />
                <button
                    type="button"
                    styleName="delete-btn"
                    onClick={this.removeTodo}
                    disabled={isLoading}
                    title="Remove Todo"
                >
                    âœ•
                </button>
            </form>
        )
    }
}

Todo.propTypes = {
    todo: PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        schedule: PropTypes.number.isRequired,
        lastEvent: PropTypes.string.isRequired,
    }).isRequired,
    saveTodo: PropTypes.func.isRequired,
    removeTodo: PropTypes.func.isRequired,
    today: PropTypes.string.isRequired,
    isLoading: PropTypes.bool.isRequired,
}

export default Todo
