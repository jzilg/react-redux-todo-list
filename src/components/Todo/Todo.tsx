import React from 'react'
import TodoType from '../../interfaces/todo.interface'
import equal from 'deep-equal'
import Icon from '../Icon'
import Urgency from '../Urgency'
import { getTodaysDate } from '../../utils/helper'
import style from './todo.scss'

interface TodoProps {
    todo: TodoType
    saveTodo: Function,
    removeTodo: Function,
    isLoading: boolean,
}

interface TodoState extends TodoType {}

class Todo extends React.Component<TodoProps, TodoState> {
    today = getTodaysDate()

    constructor(props) {
        super(props)

        const { todo } = props

        this.state = {
            id: todo.id,
            name: todo.name,
            schedule: parseInt(todo.schedule, 10),
            lastEvent: todo.lastEvent || this.today,
        }

        this.inputChange = this.inputChange.bind(this)
        this.saveTodo = this.saveTodo.bind(this)
        this.removeTodo = this.removeTodo.bind(this)
        this.setInputToday = this.setInputToday.bind(this)
    }

    setInputToday() {
        const { state } = this
        if (state.lastEvent === this.today) {
            return
        }

        this.setState({
            lastEvent: this.today,
        }, () => {
            this.saveTodo()
        })
    }

    /**
     * @returns {boolean}
     */
    todoHasChanged() {
        const { todo } = this.props
        return !equal(todo, this.state)
    }

    /**
     * @param {Object} event
     */
    inputChange(event) {
        const { name } = event.target
        let { value } = event.target

        if (name === 'schedule') {
            value = parseInt(value, 10)
        }

        this.setState({
            [name]: value,
        })
    }

    saveTodo() {
        const { saveTodo } = this.props
        saveTodo(this.state)
    }

    removeTodo() {
        const { removeTodo } = this.props
        removeTodo(this.state)
    }

    render() {
        const { isLoading } = this.props
        const { name, schedule, lastEvent } = this.state
        const saveBtnIsDisabled = () => !this.todoHasChanged() || isLoading
        const saveBtnTitle = saveBtnIsDisabled() ? '' : 'Save Todo'

        return (
            <form className={style.todo} title={name}>
                <input
                    type="text"
                    name="name"
                    value={name}
                    placeholder="Name"
                    autoComplete="off"
                    onChange={this.inputChange}
                    autoFocus={!name.length}
                    className={style.input}
                />
                <div>
                    <span>
                        Every
                    </span>
                    <input
                        type="number"
                        name="schedule"
                        className={style['schedule-input']}
                        value={schedule}
                        min="1"
                        autoComplete="off"
                        onChange={this.inputChange}
                    />
                    <span>
                        {schedule === 1 ? 'Day' : 'Days'}
                    </span>
                </div>
                <div>
                    <span>
                        Last done on
                    </span>
                    <input
                        type="date"
                        name="lastEvent"
                        className={style['last-event-input']}
                        value={lastEvent}
                        onChange={this.inputChange}
                    />
                    <button
                        type="button"
                        onClick={this.setInputToday}
                    >
                        Today
                    </button>
                </div>
                <Urgency
                    today={this.today}
                    lastEvent={lastEvent}
                    schedule={schedule}
                />
                <ul className={style['btn-list']}>
                    <li className={style['btn-list-item']}>
                        <button
                            type="button"
                            className={style['save-btn']}
                            onClick={this.saveTodo}
                            disabled={saveBtnIsDisabled()}
                            title={saveBtnTitle}
                        >
                            <Icon name="save" width="1em" height="auto" />
                        </button>
                    </li>
                    <li className={style['btn-list-item']}>
                        <button
                            type="button"
                            className={style['delete-btn']}
                            onClick={this.removeTodo}
                            disabled={isLoading}
                            title="Remove Todo"
                        >
                            âœ•
                        </button>
                    </li>
                </ul>
            </form>
        )
    }
}

export default Todo
