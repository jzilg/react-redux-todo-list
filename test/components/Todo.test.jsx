import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { shallowToJson } from 'enzyme-to-json'
import expect from 'expect'
import Todo from '../../src/components/Todo/Todo'
import { getTodaysDate } from '../../src/utils/helper'

Enzyme.configure({ adapter: new Adapter() })

describe('Todo', () => {
    const removeTodo = jest.fn()
    const saveTodo = jest.fn()
    const today = '2018-06-02'
    const todo = {
        id: 1,
        name: 'Papa anrufen',
        schedule: 4,
        lastEvent: '2018-06-03',
    }

    it('should render correctly one todo', () => {
        const component = shallow((
            <Todo
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        const tree = shallowToJson(component)
        expect(tree).toMatchSnapshot()
    })

    it('should call saveTodo', () => {
        const component = shallow((
            <Todo
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        component.instance().saveTodo()
        expect(saveTodo.mock.calls.length).toBe(1)
    })

    it('should call removeTodo', () => {
        const component = shallow((
            <Todo
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        component.instance().removeTodo()
        expect(removeTodo.mock.calls.length).toBe(1)
    })

    it('should update state if inputChanged is called', () => {
        const component = shallow((
            <Todo
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        const event = {
            target: {
                name: 'foo',
                value: 'bar',
            },
        }

        component.instance().inputChange(event)
        const newState = component.state('foo')
        expect(newState).toBe('bar')
    })

    it('should update state with number if inputChanged is called and name is schedule', () => {
        const component = shallow((
            <Todo
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))
        const event = {
            target: {
                name: 'schedule',
                value: '1',
            },
        }

        component.instance().inputChange(event)
        const newState = component.state('schedule')
        expect(newState).toBe(1)
    })

    it('should not set state.lastEvent if setInputToday is called but last event is today', () => {
        const todoWithLastEventToday = {
            ...todo,
            lastEvent: today,
        }
        const component = shallow((
            <Todo
                todo={todoWithLastEventToday}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        const instance = component.instance()
        instance.today = today
        instance.setInputToday()

        const stateToday = component.state('lastEvent')
        expect(stateToday).toBe(today)
    })

    it('should set state.lastEvent to today, if setInputToday is called and last event was in the past', () => {
        const component = shallow((
            <Todo
                todo={todo}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        const instance = component.instance()
        instance.today = today
        instance.setInputToday()

        const stateToday = component.state('lastEvent')
        expect(stateToday).toBe(today)
    })

    it('should set initial state.lastEvent to today if none is set', () => {
        const todoWithNoLastEvent = {
            ...todo,
            lastEvent: '',
        }
        const component = shallow((
            <Todo
                todo={todoWithNoLastEvent}
                isLoading={false}
                removeTodo={removeTodo}
                saveTodo={saveTodo}
            />
        ))

        const state = component.state('lastEvent')
        expect(state).toBe(getTodaysDate())
    })
})
