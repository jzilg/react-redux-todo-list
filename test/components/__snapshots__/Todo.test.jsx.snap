// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should render correctly one todo 1`] = `
<form
  key="1"
  styleName="todo"
  title="Papa anrufen"
>
  <input
    autoComplete="off"
    autoFocus={false}
    name="name"
    onChange={[Function]}
    placeholder="Name"
    styleName="input"
    type="text"
    value="Papa anrufen"
  />
  <div>
    <span>
      Every
    </span>
    <input
      autoComplete="off"
      min="1"
      name="schedule"
      onChange={[Function]}
      styleName="schedule-input"
      type="number"
      value={4}
    />
    <span>
      Days
    </span>
  </div>
  <div>
    <span>
      Last done on
    </span>
    <input
      name="lastEvent"
      onChange={[Function]}
      styleName="last-event-input"
      type="date"
      value="2018-06-03"
    />
    <button
      onClick={[Function]}
      type="button"
    >
      Today
    </button>
  </div>
  <Urgency
    lastEvent="2018-06-03"
    schedule={4}
    today="2018-06-02"
  />
  <ul
    styleName="btn-list"
  >
    <li
      styleName="btn-list-item"
    >
      <button
        disabled={true}
        onClick={[Function]}
        styleName="save-btn"
        title=""
        type="button"
      >
        <Icon
          className=""
          height="auto"
          name="save"
          width="1em"
        />
      </button>
    </li>
    <li
      styleName="btn-list-item"
    >
      <button
        disabled={false}
        onClick={[Function]}
        styleName="delete-btn"
        title="Remove Todo"
        type="button"
      >
        âœ•
      </button>
    </li>
  </ul>
</form>
`;
